front: GEV: display power restriction on GEV

Yohh opened the pull request:
close #3805

Yohh committed:
front: GEV: display power restriction on GEV; add power restriction setting in SpeedSpaceSettings; update simulation.json

- add condition if used and/or seen restriction is undefined

- fix comments

Castavo commented on front/src/applications/operationalStudies/components/SimulationResults/SpeedSpaceChart/SpeedSpaceChart.jsx:
```diff
@@ -257,8 +261,27 @@ export default function SpeedSpaceChart(props) {
+                padding: '0.2rem',
+              }}
+            >{`${t('speedSpaceSettings.powerRestriction')}: ${
+              trainSimulation.electrificationConditions[0].used_restriction
```
The first `used_restriction` might be `null` if the user entered a power restriction code which is valid 25000 electrification and the first electrification encountered is 1500

Yohh replied:
done

clarani commented on front/src/applications/operationalStudies/components/SimulationResults/SpeedSpaceChart/SpeedSpaceChart.jsx:
```diff
@@ -189,6 +195,31 @@ export default function SpeedSpaceChart(props) {
+      });
+      if (!restrictionPower) {
+        setRestrictionPower(
+          `${t('speedSpaceSettings.powerRestriction')} ${t('powerRestriction.unknown')}`
```
```suggestion
          ${t('powerRestriction.unknown')}`
```
Only use 1 traduction (bad practice, it will be broken if the speedSpaceSettings.powerRestriction is modified !). You should modify powerRestriction.unknown: "Restriction de puissance inconnue"

Yohh replied:
done

clarani commented on front/src/applications/operationalStudies/components/SimulationResults/SpeedSpaceChart/SpeedSpaceChart.jsx:
```diff
@@ -57,6 +61,8 @@ export default function SpeedSpaceChart(props) {
   const [rotate, setRotate] = useState(false);
   const [showSettings, setShowSettings] = useState(false);
 
+  const [restrictionPower, setRestrictionPower] = useState('');
```
Can you move it above ? (alphabetical order for the state vars)

Yohh replied:
done

clarani commented on front/src/applications/operationalStudies/components/SimulationResults/SpeedSpaceChart/SpeedSpaceChart.jsx:
```diff
@@ -189,6 +195,31 @@ export default function SpeedSpaceChart(props) {
+  useEffect(() => {
+    setRestrictionPower('');
+    trainSimulation.electrificationConditions.forEach((elem) => {
+      if (elem.used_restriction)
```
I don't see used_restriction in any type in the code ? We should try to type everything

Yohh replied:
done

clarani committed:
fix bug

Yohh committed:
front: GEV: create drawPowerRestriction.jsx, remove related component from SpeedSpaceChart.jsx; add condition in d3helpers.jsx; edit consts.ts with power restrictions data

- group rects by restriction; add striped pattern

- edit text when incompatible restriction

anisometropie commented on front/src/applications/operationalStudies/consts.ts:
```diff
@@ -322,18 +332,75 @@ export const createProfileSegment = (
+  segment.height_end = 24;
+  segment.height_middle = (segment.height_start + segment.height_end) / 2;
+  segment.seenRestriction = electrificationConditions.seen_restriction || '';
+  segment.usedRestriction = electrificationConditions.used_restriction || '';
```
```suggestion
  const segment: PowerRestrictionSegment = {
    position_start: electrificationConditions.start,
    position_end: electrificationConditions.stop,
    position_middle: (electrificationConditions.start + electrificationConditions.stop) / 2,
    lastPosition: fullElectrificationConditions.slice(-1)[0].stop,
    height_start: 4,
    height_end: 24,
    height_middle: (segment.height_start + segment.height_end) / 2,
    seenRestriction: electrificationConditions.seen_restriction || '',
    usedRestriction: electrificationConditions.used_restriction || '',
    isStriped: false,
    isRestriction: false,
    isIncompatiblePowerRestriction: false,
  };
```

Yohh replied:
done
